language/vm/vm_runtime/src/interpreter.rs

```diff
@@ -0,0 +1,1204 @@
+    ) -> VMResult<()> {
+        let account_struct_id = account_module
+            .struct_defs_table
+            .get(&*ACCOUNT_STRUCT_NAME)
+            .ok_or_else(|| VMStatus::new(StatusCode::LINKER_ERROR))?;
+        let account_struct_def = self
+            .module_cache
+            .resolve_struct_def(account_module, *account_struct_id, &self.gas_meter)?
+            .ok_or_else(|| VMStatus::new(StatusCode::LINKER_ERROR))?;
+
+        let account_resource = self.operand_stack.pop_as::<Struct>()?;
+        // TODO: Adding the freshly created account's expiration date to the TransactionOutput here.
+        let account_path = Self::make_access_path(account_module, *account_struct_id, addr);
+        self.data_view
+            .move_resource_to(&account_path, account_struct_def, account_resource)
+    }
+
+    /// Create an account on the blockchain by calling into `CREATE_ACCOUNT_NAME` function stored
+    /// in the `ACCOUNT_MODULE` on chain.
+    // REVIEW: this should not live here
+    pub fn create_account_entry(&mut self, addr: AccountAddress) -> VMResult<()> {
+        let account_module = self
+            .module_cache
+            .get_loaded_module(&ACCOUNT_MODULE)?
+            .ok_or_else(|| VMStatus::new(StatusCode::LINKER_ERROR))?;
+
+        // TODO: Currently the event counter will cause the gas cost for create account be flexible.
+        //       We either need to fix the gas stability test cases in tests or we need to come up
+        //       with some better ideas for the event counter creation.
+        self.gas_meter.disable_metering();
+        // Address will be used as the initial authentication key.
+        self.execute_function(
+            &ACCOUNT_MODULE,
+            &CREATE_ACCOUNT_NAME,
+            vec![Value::byte_array(ByteArray::new(addr.to_vec()))],
+        )?;
+        self.gas_meter.enable_metering();
+
+        self.save_account(account_module, addr)
+    }
+
+    //
+    // Debugging and logging helpers.
+    //
+
+    /// Given an `VMStatus` generate a core dump if the error is an `InvariantViolation`.
+    fn maybe_core_dump(
+        &self,
+        err: VMStatus,
+        current_frame: &Frame<'txn, FunctionRef<'txn>>,
+    ) -> VMStatus {
+        if err.is(StatusType::InvariantViolation) {
+            crit!(
+                "Error: {:?}\nCORE DUMP: >>>>>>>>>>>>\n{}\n<<<<<<<<<<<<\n",
+                err,
+                self.get_internal_state(current_frame),
+            );
+        }
+        err
+    }
+
+    /// Generate a string which is the status of the interpreter: call stack, current bytecode
+    /// stream, locals and operand stack.
+    ///
+    /// It is used when generating a core dump but can be used for debugging of the interpreter.
+    /// It will be exposed via a debug module to give developers a way to print the internals
+    /// of an execution.
+    fn get_internal_state(&self, current_frame: &Frame<'txn, FunctionRef<'txn>>) -> String {
```

Comment:
Annotate these two functions with `#[cfg(...)]` so they can't be used in prod by mistake? (No strong opinion on this and not sure how practical it is though.)

Reply:
well those 2 functions **are used** in production! :)
I'd say that is the best part of this diff that I did not advertise well enough.
If there is an invariant violation we now dump the entire state of the interpreter in the log. That is pretty important for us to debug the system. And we have a reasonably good printing already here (I know I should have put a preview in the description but...)
Moreover very soon we will have an implementation of a Debug module that will offer this API as something callable from Move to debug your code.

Reply:
Could that be too risky? I know that in web apps it is bad to dump out internal state because it can hint at security issues.